In [1]:
import pandas as pd
import numpy as np

We have a dataset of used cars sold in eBay Germany, we want to clean the dataset and analyze the included used cars listing

In [ ]:
autos=pd.read_csv("autos.csv", encoding="Latin-1")

In [130]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34306 entries, 0 to 49999
Data columns (total 18 columns):
date_crawled          34306 non-null int64
name                  34306 non-null object
seller                34306 non-null object
offer_type            34306 non-null object
price_usd             34306 non-null float64
abtest                34306 non-null object
vehicle_type          34306 non-null object
registration_year     34306 non-null int64
gearbox               34306 non-null object
power_ps              34306 non-null int64
model                 34306 non-null object
odometer_km           34306 non-null int64
registration_month    34306 non-null int64
fuel_type             34306 non-null object
brand                 34306 non-null object
unrepaired_damage     34306 non-null object
ad_created            34306 non-null int64
last_seen             34306 non-null int64
dtypes: float64(1), int64(7), object(10)
memory usage: 5.0+ MB


In [127]:
autos

date_crawled                                               name  \
0          20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1          20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2          20160326                         Volkswagen_Golf_1.6_United   
3          20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4          20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
7          20160316                               Golf_IV_1.9_TDI_90PS   
9          20160316          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
12         20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13         20160323                                 Audi_A3_1.6_tuning   
16         20160316                                Opel_Vectra_B_Kombi   
17         20160329                          Volkswagen_Scirocco_2_G60   
19         20160317        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
21         20160306                      Porsche_911_Carrera_4S_Cabrio   
22         20160328                               MINI_Cooper_S_Cabrio   
24         20160403                         BMW_535i_xDrive_Sport_Aut.   
26         20160403                                Volkswagen_Polo_Fox   
28         20160319                                      MINI_Cooper_D   
29         20160402  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31         20160314                          Noch_gut_erhaltenen_C_320   
33         20160315  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
35         20160329                                        Ford_Mondeo   
36         20160316          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37         20160323                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38         20160321                        Volkswagen_Golf_1.4_Special   
39         20160311                 Mercedes_C_180_Kombi__zu_verkaufen   
40         20160307  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
43         20160308           Audi_A1_1.2_TFSI_Attraction_mit_Garantie   
44         20160312                                       Zu_verkaufen   
45         20160312                                 Toyota_Starlet_1.3   
47         20160328                                Mercedes_Benz_A_150   
...             ...                                                ...   
49953      20160330                                       Ford_Mustang   
49956      20160328        Toyota_Yaris_1.3_VVT_i_Executive_MODEL_2007   
49959      20160328           Mercedes_Benz_C_180_T_Kompressor_Classic   
49962      20160314  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49963      20160326            Mercedes_Benz_B_200_CDI_Special_Edition   
49965      20160311          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966      20160402  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49968      20160401                  Mercedes_Benz_190_D_2.5_Automatik   
49969      20160317             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970      20160321  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971      20160329                                         W.Lupo_1.0   
49972      20160326          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973      20160327                   Mercedes_Benz_SLK_200_Kompressor   
49975      20160327          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976      20160319  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977      20160331                        Mercedes_Benz_C200_Cdi_W203   
49978      20160404                        Mercedes_Benz_E_200_Classic   
49979      20160320                      Volkswagen_Polo_1.6_TDI_Style   
49981      20160315             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49986      20160404  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987      20160322  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...   
49988      20160328                                         BMW_330_Ci   
49990      

In [ ]:
autos.isnull().sum()

There are multiple columns with null values, in particular the following columns: vehicleType (5095), gearbox(2680), model(2758), fuelType (4482), notRepairedDamage (9829).

In [ ]:
corrections=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_nr', 'postal_code',
       'last_seen']
autos.columns=corrections

I just changed the column labels from Camelcase to snake case which means replacing the spaces with underscores and turning all the upper cases to lower cases.

In [ ]:
autos.describe(include='all')

-The 99.99% of the values in the columns "seller" and "offer_type" are respectively "privat" and "Angebot" so we could directly consider that all the offers were posted by privates and their type is Angebot, discarding these two columns.

-The values in the columns "price" and "odometer" are number easily convertible to int/float.

-The "registration_year/month", "power_ps", "pictures_nr" and "postal_code" columns already include integer values and not strings.

In [ ]:
# Let's change the type of datas in the Price and Odometer columns from strings to float or int
# and rename the columns in price_usd and odometer_km


autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(float)
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
#print(autos["odometer"])
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)
autos.rename({"price":"price_usd"},axis=1,inplace=True)
print(autos.dtypes)

In [ ]:
# Let's have a look at the different prices of the cars for sale to see
# how many cars are sold for the same price and let's order the count in ascending order

autos["price_usd"].value_counts().sort_index(ascending=True).head(100)

In [ ]:
# Let's consider only cars sold for a price included in the range [200,999999]USD, discarding all the other rows.

out_range=autos.loc[~autos["price_usd"].between(200,999999)]
autos.drop(out_range.index.values,axis=0,inplace=True)

While the values in the "odometer_km" column seem to be acceptable (there's not a logic limit to set), the values in the "price_usd" column range from 0 to 10 million dollars. 
Appearing unlikely that someone would use eBay to advertise a free car or a car worth over a million dollars, we decided to exclude prices below 200$ and over 1 million dollar.
We obtained the indexes of the rows with prices outside this range thanks to the method "index.values" and then used the .drop() method to exclude them.

There are dates written in the columns "date_crawled", "ad_created", "registration_month", "registration_year" and "last_seen". We could extract the digits of the different dates using the str.split() method and create new columns for the years and the months.

In [ ]:
#autos["date_crawled"].value_counts().head()
#autos["ad_created"].value_counts(normalize=True).head()
autos["last_seen"].value_counts().head()


Both the date_crawled and the last_seen values are included in the range 05/03/2016 and 07/04/2016, while the ad_created values belong to the range 11/06/2015 - 07/04/2016. The most recurring dates belonging to these 3 columns are 03-04-2016 (ad_created, #:1857 ~4%), 07-04-2016 (last_seen, #:8 ~0.017%) and multiple dates for date_crawled (all in quantity of #:3,which means approximately 0.0063%).

In [ ]:
autos["registration_year"].describe()

It appears evident that many outliers are included in the "registration_year" column. Having a look at the values of the column through the "describe" method, it appears peculiar to notice that the minimum value is 1000 and the maximum is 9999 which are, obviously, impossible values to accept.

In [ ]:
out_of_range_years=autos.loc[~autos["registration_year"].between(1900,2016)]
autos.drop(out_of_range_years.index.values,axis=0,inplace=True)

In [ ]:
autos["registration_year"].value_counts()

We decided to count the number of listings with cars that fall outside the 1900 - 2016 and to cancel their rows.
Now it appears that 3019 (~6.5%) cars in this list were registered in the year 2000.

In [ ]:
# Let's calculate the mean price per brand sold on eBay

first_brands=autos["brand"].unique()

mean_per_brand={}
for brand in first_brands:
    df_branded=autos.loc[autos["brand"]==brand]
    mean_per_brand[brand]=df_branded["price_usd"].mean()
print(mean_per_brand)

In [123]:
# We repeated the calculation just for the 10 most sold brands, first creating a DataFrame 
# with the 10 most sold brands (first_brand) and then creating a dictionary (dictionary) 
# to associate a mean price to every brand

first_brand=autos["brand"].value_counts(normalize=True).head(10)
print(first_brand)

dictionary={}
for index in first_brand.index:
    red_df=autos.loc[autos["brand"]==index]
    mean_br=red_df["price_usd"].mean()
    dictionary[index]=mean_br

print(dictionary)

volkswagen       0.206145
bmw              0.117356
mercedes_benz    0.103422
opel             0.101265
audi             0.092200
ford             0.066519
renault          0.044569
peugeot          0.029907
fiat             0.023261
seat             0.018889
Name: brand, dtype: float64
{'bmw': 9237.371087928464, 'volkswagen': 6489.714649321267, 'ford': 4872.069237510956, 'renault': 2968.1294964028775, 'seat': 5013.733024691358, 'mercedes_benz': 9556.609075535513, 'peugeot': 3490.293372319688, 'fiat': 3350.2406015037595, 'opel': 3495.1171560161197, 'audi': 10591.7094530509}


Having analyzed the first ten most occurring brands (until the ~2% of the total), we can say that their mean price varies in the range [2552-9406]USD. With the renault cars being generally the cheapest ones and the audi cars the most expensive ones.
The most expensive brands are:
1)Audi 9406USD - Occurrence rate ~9%
2)Mercedes Benz 8691USD - Occurrence rate ~10%
3)BMW 8402USD - Occurrence rate ~11%
SO we can state that ~30% of the used cars sold in eBay Germany are german expensive cars and include cars of just 3 three brands. While putting together the three cheapest brands (renault(2552USD), fiat(2925USD) and opel (3077USD)) we are taking into account just ~17% of the total.
This could be a way to measure the wealth of a country, Germany is the economical leading country of Europe, its population is generally wealthier than the average of people living in the same continent and therefore they tend to buy more expensive goods.

In [ ]:
# Let's create 3 new dictionaries associating to each brand its mean ratio power/price, its mean power and its mean odometer

mean_power={}
mileage={}
power_on_price={}
for index in first_brand.index:
    reduced=autos.loc[autos["brand"]==index]
    power=reduced["power_ps"].mean()
    odom=reduced["odometer_km"].mean()
    mean_pr=reduced["price_usd"].mean()
    pop=power/mean_pr
    power_on_price[index]=pop
    mean_power[index]=power
    mileage[index]=odom
#print(mean_power)
#print(mileage)
#print(power_on_price)

In [124]:
# Let's convert the newly created dictionaries into pandas Series and then let's make a new DataFrame with all the 
# calculated mean values associated to the most popular brands

mean_series_1=pd.Series(dictionary)
mean_series_2=pd.Series(mean_power)
mean_series_3=pd.Series(mileage)
mean_series_4=pd.Series(power_on_price)

mean_df=pd.DataFrame(mean_series_1, columns=['mean_price_usd'])
mean_df["mean_power_ps"]=mean_series_2
mean_df["mean_mileage_km"]=mean_series_3
mean_df["power_on_price"]=mean_series_4
mean_df

mean_price_usd  mean_power_ps  mean_mileage_km  power_on_price
audi              9406.093244     161.527300    129260.782847        0.017173
bmw               8402.666078     168.088536    132792.500982        0.020004
fiat              2925.954743      71.299391    116949.521323        0.024368
ford              4199.720140     103.312897    124096.251588        0.024600
mercedes_benz     8691.717450     152.876510    131091.722595        0.017589
opel              3077.577232      97.728604    129231.800371        0.031755
peugeot           3142.016047      92.992706    126929.248724        0.029597
renault           2552.515251      82.334115    128052.557485        0.032256
seat              4505.462740      98.310096    121604.567308        0.021820
volkswagen        5712.041141     100.064606    128779.201985        0.017518

Having said about the mean prices, we analyzed also the mean power and it seems that the most powerful cars sold in eBay Germany are also the most expensive ones: Audi (161.5), BMW (168.1) and Mercedes (152.9). These brands occupy the first spots also in the "most used cars" chart, infact the mean mileage for these brands are the highest ones (exceeding or almost exceeding the 130000 kms), curious fact, the least used is the fiat with a mean mileage of 116000 kms.
Finally we analyzed the "quality on price" ratio or, it may be better to call it, the "power on price" ratio and it seems that the most convenient car to buy for an eBay Germany client is a Renault (0.032), immediately followed by Opel (0.031) while the most expensive brands (Audi, Mercedes and BMW) and the Volkswagen seem to be the least convenient ones [0.017-0.02]. So, in conclusion, Renault and Opel cars are twice as convenient, for a buyer, than the most expensive cars.

In [ ]:
# Let's translate all the terms in german to english

unique_seller=autos["seller"].unique()
autos["seller"]=autos["seller"].str.replace("privat","private")
autos["seller"]=autos["seller"].str.replace("privatee","private")

unique_offer=autos["offer_type"].unique()
autos["offer_type"]=autos["offer_type"].str.replace("Angebot","offer")

unique_abtest=autos["abtest"].unique()

unique_veh_type=autos["vehicle_type"].unique()
autos["vehicle_type"]=autos["vehicle_type"].str.replace("kleinwagen","city car")
autos["vehicle_type"]=autos["vehicle_type"].str.replace("kombi","station wagon")
autos["vehicle_type"]=autos["vehicle_type"].str.replace("andere","other")

unique_gear=autos["gearbox"].unique()
autos["gearbox"]=autos["gearbox"].str.replace("manuell","manual")
autos["gearbox"]=autos["gearbox"].str.replace("automatik","automatic")

unique_fuel_type=autos["fuel_type"].unique()
autos["fuel_type"]=autos["fuel_type"].str.replace("benzin","gasoline")
autos["fuel_type"]=autos["fuel_type"].str.replace("cng","natural gas")
autos["fuel_type"]=autos["fuel_type"].str.replace("elektro","electric")
autos["fuel_type"]=autos["fuel_type"].str.replace("andere","other")

unique_damage=autos["unrepaired_damage"].unique()
autos["unrepaired_damage"]=autos["unrepaired_damage"].str.replace("ja","yes")
autos["unrepaired_damage"]=autos["unrepaired_damage"].str.replace("nein","no")

autos["model"]=autos["model"].str.replace("andere","other")

In [ ]:
# Let's get rid of all the rows with null values

column_labels=autos.columns
for label in column_labels:
    exclude_nan=autos.loc[autos[label].isnull()]
    autos.drop(exclude_nan.index.values, axis=0, inplace=True)

Getting rid of all the rows with null values, we reduced the number of rows from 50.000 to 34.306. Which means that around 31% of the datas were discarded.

In [ ]:
# Let's separate the dates in the format YYYY-MM-DD from the time. We are using str.split that will replace the
# initial column with the first column after the separation took place

autos["date_crawled"]=autos["date_crawled"].str.split(n=1, expand=True)
autos["date_crawled"]=autos["date_crawled"].str.replace("-","").astype(int)

autos["ad_created"]=autos["ad_created"].str.split(n=1, expand=True)
autos["ad_created"]=autos["ad_created"].str.replace("-","").astype(int)

autos["last_seen"]=autos["last_seen"].str.split(n=1, expand=True)
autos["last_seen"]=autos["last_seen"].str.replace("-","").astype(int)

In [ ]:
# We are deleting the "pictures_nr" column (always=0) and the "postal_code" which seems irrelevant

autos.drop("pictures_nr", axis=1, inplace=True)
autos.drop("postal_code", axis=1, inplace=True)

In [118]:
# Checking if the names of the cars' models are correctly
# written in english



data_volks=autos.loc[autos["brand"]=="volkswagen"]
model_volks=data_volks["model"].unique()
#print(model_volks)

autos["model"]=autos["model"].str.replace("reihe","series")
data_bmw=autos.loc[autos["brand"]=="bmw"]
model_bmw=data_bmw["model"].unique()
#print(model_bmw)

autos["model"]=autos["model"].str.replace("klasse","class")
data_merc=autos.loc[autos["brand"]=="mercedes_benz"]
model_merc=data_merc["model"].unique()
#print(model_merc)

data_opel=autos.loc[autos["brand"]=="opel"]
model_opel=data_opel["model"].unique()
#print(model_opel)

data_audi=autos.loc[autos["brand"]=="audi"]
model_audi=data_audi["model"].unique()
#print(model_audi)


# Making lists of the most common models of the most common brands

comm_brands=autos["brand"].value_counts().head()

for brand in comm_brands.index:
    datas_per_brand=autos.loc[autos["brand"]==brand]
    if brand=="volkswagen":
        volks_mod=datas_per_brand["model"].value_counts().head()
    if brand=="bmw":
        bmw_mod=datas_per_brand["model"].value_counts().head()
    if brand=="mercedes_benz":
        merc_mod=datas_per_brand["model"].value_counts().head()
    if brand=="opel":
        opel_mod=datas_per_brand["model"].value_counts().head()
    if brand=="audi":
        audi_mod=datas_per_brand["model"].value_counts().head()
print(audi_mod)

a4       965
a6       661
a3       643
other    207
tt       121
Name: model, dtype: int64


In [119]:
# Creating dictionaries for the most commond models and their counts

volks={}
count_volks={}
a=1
for i in volks_mod.index:
    volks[a]=i
    a+=1
a=1
for i in volks_mod:
    count_volks[a]=i
    a+=1

bmw={}
count_bmw={}
a=1
for i in bmw_mod.index:
    bmw[a]=i
    a+=1
a=1
for i in bmw_mod:
    count_bmw[a]=i
    a+=1
    
merc={}
count_merc={}
a=1
for i in merc_mod.index:
    merc[a]=i
    a+=1
a=1
for i in merc_mod:
    count_merc[a]=i
    a+=1
    
opel={}
count_opel={}
a=1
for i in opel_mod.index:
    opel[a]=i
    a+=1
a=1
for i in opel_mod:
    count_opel[a]=i
    a+=1
    
audi={}
count_audi={}
a=1
for i in audi_mod.index:
    audi[a]=i
    a+=1
a=1
for i in audi_mod:
    count_audi[a]=i
    a+=1
    
print(audi, count_audi)

{1: 'a4', 2: 'a6', 3: 'a3', 4: 'other', 5: 'tt'} {1: 965, 2: 661, 3: 643, 4: 207, 5: 121}


In [120]:
# Transforming dictionaries into series and merging them into a DataFrame (most_common_mod)

s1=pd.Series(volks)
s2=pd.Series(count_volks)
s3=pd.Series(merc)
s4=pd.Series(count_merc)
s5=pd.Series(bmw)
s6=pd.Series(count_bmw)
s7=pd.Series(opel)
s8=pd.Series(count_opel)
s9=pd.Series(audi)
s10=pd.Series(count_audi)

In [121]:
most_common_mod=pd.DataFrame(s1, columns=["volkswagen"])
most_common_mod["count_volkswagen"]=s2
most_common_mod["mercedes_benz"]=s3
most_common_mod["count_mercedes_benz"]=s4
most_common_mod["bmw"]=s5
most_common_mod["count_bmw"]=s6
most_common_mod["opel"]=s7
most_common_mod["count_opel"]=s8
most_common_mod["audi"]=s9
most_common_mod["count_audi"]=s10

In [122]:
most_common_mod

volkswagen  count_volkswagen mercedes_benz  count_mercedes_benz       bmw  \
1         golf              2728       c_class                  933       3er   
2         polo              1103       e_class                  763       5er   
3       passat              1022       a_class                  415       1er   
4  transporter               532         other                  380  x_series   
5       touran               333           clk                  193  z_series   

   count_bmw    opel  count_opel   audi  count_audi  
1       2074   corsa        1057     a4         965  
2        921   astra        1016     a6         661  
3        459  vectra         366     a3         643  
4        269  zafira         307  other         207  
5        109   other         116     tt         121

In [126]:
# Calculating the mean prices of cars having splitted the
# mileage in three different parts. x<50000km, 50000<=x<10000 and x>=100000

low_mileage=autos.loc[autos["odometer_km"]<50000]
mid_mileage=autos.loc[(autos["odometer_km"]>=50000)&(autos["odometer_km"]<100000)]
high_mileage=autos.loc[autos["odometer_km"]>=100000]

me_pr_lo_mi=low_mileage["price_usd"].mean()
me_pr_mi_mi=mid_mileage["price_usd"].mean()
me_pr_hi_mi=high_mileage["price_usd"].mean()

print(me_pr_lo_mi, me_pr_mi_mi, me_pr_hi_mi)

17873.231374172185 11417.061060123619 5017.191292230048


It appears evident that cars with a lower mileage (inferior to 50.000 kms) have a higher mean cost (17873 USD) which is 3.56 times the cost of cars with the highest mileage (superior to 100.000 kms)(5017 USD). Cars with a medium mileage (between 50.000 and 100.000 kms) have a cost wich is 64% of the cost of cars with the lowest mileage (11417 USD).

In [129]:
# Calculating the mean prices of cars with repaired/unrepaired damages

dam_cars=autos.loc[autos["unrepaired_damage"]=="yes"]
non_dam_cars=autos.loc[autos["unrepaired_damage"]=="no"]

me_pr_da_ca=dam_cars["price_usd"].mean()
me_pr_no_da_ca=non_dam_cars["price_usd"].mean()

print(me_pr_da_ca, me_pr_no_da_ca)

2812.210153148043 7388.9866471734895


Even more evidently, damaged (and never repaired) cars have a mean cost which is 38% of the cost of never damaged or damaged and repaired cars (2812.21 USD vs. 7388.98 USD).
So damaged cars are 62% cheaper than the undamaged cars.